smk/gene_annotation.smk

```diff
@@ -182,79 +182,112 @@ rule prokka_for_genome:
+                                $F[7] = "CRISPR";
+                            }}
+                            $F[9] = sprintf("%s_%s_%s", $F[0], $F[1]+1, $F[2]);
+                        }}' \
+                    -e '$F[1]++;' \
+                    -e 'print join("\\t", @F);' \
                 > out.bed
-            rsync -a out.bed {output.bed}
-        ) >& {log}
-        """
 
-# Combine multiple GFF/BED entries for the same entity - e.g., gene/CDS, gene/tRNA
-rule merge_features_in_locus:
-    input:
-        bed = rules.make_bed_for_genome.output.bed,
-        faa = rules.rename_prokka_output_faa.output.faa
-    output:
-        bed="{wd}/DB/{post_analysis_dir}/3-merged-features/{genome}.bed",
-        faa="{wd}/DB/{post_analysis_dir}/3-merged-features/{genome}.faa"
-    log:
-        "{wd}/logs/DB/{post_analysis_dir}/3-merged-features/{genome}.merge_bed_items.log"
-    resources:
-        mem=5
-    threads: 1
-    conda:
-        config["minto_dir"]+"/envs/r_pkgs.yml"
-    shell:
-        """
-        time (
-            Rscript {script_dir}/merge_features_in_locus.R {input.bed} {input.faa} {output.bed} {output.faa}
+            # Save output files
+            rsync -a out.bed {output.bed}
+            rsync -a out.fna {output.fna}
+            rsync -a out.faa {output.faa}
         ) >& {log}
         """
 
+# Get a list of genome bed files
 def get_genome_bed(wildcards):
     #Collect the BED files for MAGs
-    genomes = get_genomes_from_refdir(reference_dir)
-    result = expand("{wd}/DB/{post_analysis_dir}/3-merged-features/{genome}.bed",
+    result = expand("{wd}/DB/{post_analysis_dir}/2-postprocessed/{genome}.bed",
                     wd=wildcards.wd,
                     post_analysis_dir=wildcards.post_analysis_dir,
                     genome=genomes)
     return(result)
 
+# Replace 'score' in 5th column with 'length' = stop - start + 1.
+# Retain only first word in ID column
+# Write full info into full BED file.
+# Write only chr,start,stop,length,ID in mini BED file.
+
+def process_genome_bed_list(input_list, output_full, output_mini, log_file):
+    import pandas as pd
+    import datetime
+
+    def logme(stream, msg):
+        print(datetime.datetime.now().strftime("%d/%m/%Y %H:%M:%S"), msg, file=stream)
+
+    colnames = ['chr', 'start', 'stop', 'name', 'length', 'strand', 'source', 'feature', 'frame', 'ID']
+    drop_cols = ['name', 'strand', 'source', 'feature', 'frame']
+    with open(str(log_file), 'w') as f:
+        logme(f, "INFO: reading file 0")
+
+        df = pd.read_csv(input_list[0], header=None, names=colnames, sep="\t", memory_map=True)
+        df['length'] = df['stop'] - df['start'] + 1
+        df = df.replace(to_replace={'ID': r'[;\s].*'}, value='', regex=True)
+        df.to_csv(output_full, sep="\t", index=False, header=True)
+
+        df = df.drop(drop_cols, axis='columns')
+        df.to_csv(output_mini, sep="\t", index=False, header=False)
+
```

Comment:
If I perceive this right, and the 0. only differs in having a header in the full bed, then one could make it simpler by just writing the header first, and then start the loop with the appends from 0.

Reply:
Absolutely right. Will fix it.

Reply:
Simplified the function. Much more straightforward now, thanks.